### PR TITLE
fix(check): replace jsonpath-plus expression with traversal for dupe fields

### LINKED ISSUE #108: replace JSON path expression in mustNotHaveDuplicateFieldNamesAfterEvaluation with a traversal
**Describe the bug**
The JSON path expression used to identify the objects in the payload to check for duplicate fields throws an error when it encounters an array. This looks to be an issue in the jsonpath-plus library.

**To Reproduce**
Use the input for #104 (part of the repo now in `valid-parameters-issue104.json` and remove the try/catch added in `mustNotHaveDuplicateFieldNamesAfterEvaluation`

**Expected behavior**
Shouldn't throw a runtime exception from the jsonpath-plus library. Should instead correctly report the payload template as valid or invalid.

**Version:**
3.0.x

**Additional context**
I think the fix is to replace the jsonpath-plus specific expression `$..[Parameters,ResultSelector]..*[?(@property.match(/^.+\.\$$/))][0]^^` with a simple graph traversal. There's too much dark magic in the expression evaluation even without this array handling issue.

### PR DESCRIPTION
**Pull request by massfords**
closes #108

### CODE REVIEW

File: src/__tests__/definitions/invalid-duplicate-fields-nested.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      "Resource": "arn:aws:lambda:::function:publish-to-slack-lambda",
+      "Parameters": {
+        "slackMessage": {
+          "channel.$": "$.slackErrorChannel",
+          "blocks": [
+            {
+              "type": "section",
+              "text": {
+                "type": "mrkdwn",
+                "type.$": "$.slackMessage.errorMessage"
```
Comment (massfords) [inline src/__tests__/definitions/invalid-duplicate-fields-nested.json:16]: these fields are in conflict

File: src/__tests__/definitions/invalid-duplicate-fields.json [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  "Comment": "Contains a duplicate field in the payload template. channel.$ and channel are in conflict.",
+  "StartAt": "Publish to Slack",
+  "States": {
+    "Publish to Slack": {
+      "Type": "Task",
+      "Resource": "arn:aws:lambda:::function:publish-to-slack-lambda",
+      "Parameters": {
+        "slackMessage": {
+          "channel.$": "$.slackErrorChannel",
+          "channel": "conflicting field",
```
Comment (massfords) [inline src/__tests__/definitions/invalid-duplicate-fields.json:11]: these field are in conflict

File: src/checks/duplicate-payload-template-fields.ts [hunk 3]
```diff
@@ -42,7 +26,12 @@ export const mustNotHaveDuplicateFieldNamesAfterEvaluation: AslChecker = (defini
                 });
             }
             keys[keyPostEval] = current ? current + 1 : 1;
+            if (typeof parent[key] === "object") {
+                inspectKeys(parent[key] as Record<string, unknown>);
+            }
```
Comment (massfords) [inline src/checks/duplicate-payload-template-fields.ts:31]: here's the traversal to visit all child nodes of the payload template

File: src/checks/duplicate-payload-template-fields.ts [hunk 4]
```diff
@@ -9,28 +9,12 @@ export const mustNotHaveDuplicateFieldNamesAfterEvaluation: AslChecker = (defini
...
-            // extension 1: @property.match enables matching on the matched property name.
-            //              This finds and selects all nodes with a name ending in .$
-            // extension 2: ^ operator twice at the end selects the parent node instead of the value for the node
-            //
-            // also, don't miss the array offset [0] before selecting the grandparent.
-            // There may be multiple fields within a node that could be in conflict.
-            // We're checking all the keys in the parent node, so we only need a single
-            // result for that parent.
-            path: `$..[Parameters,ResultSelector]..*[?(@property.match(/^.+\\.\\$$/))][0]^^`
-        })
```
Comment (massfords) [inline src/checks/duplicate-payload-template-fields.ts:27]: The expression (when it worked), identified all of the objects containing a key with a .$ suffix. This reduced the number of objects to inspect to just the ones that could be in conflict.

This expression didn't handle array types in the payload, perhaps due to a bug in jsonpath-plus. The descendants operator (..) and predicate with `property.match` throws if it encounters an array. The desired behavior IMHO would be to traverse the array's elements but that's not what the lib does.

Even if it did work, it's using extensions that aren't as widely known or documented. A simple imperative traversal like below is better than a declarative approach requiring all of these extra comments.

**Review by ChristopheBougere — APPROVED** [review]
Sounds good, thanks for the explanation!

**Pull request merged**